File: rb_101/small_problems/easy_6/exercise1.rb
DEGREE = "\xC2\xB0"
MINUTE = "'"
SECOND = "\""

MINUTES_PER_DEGREE = 60
SECONDS_PER_MINUTE = 60

def dms(degrees)
  total_minutes = (degrees * MINUTES_PER_DEGREE)
  degrees, minutes = total_minutes.divmod(MINUTES_PER_DEGREE)
  remaining_total_seconds = minutes * SECONDS_PER_MINUTE
  minutes, seconds = remaining_total_seconds.divmod(SECONDS_PER_MINUTE)

  degrees.to_s + DEGREE + format('%0.2d', minutes.round(2)) + MINUTE +
  format('%0.2d', seconds.round(2)) + SECOND
end

puts dms(30) == %(30°00'00")
puts dms(76.73) == %(76°43'48")
puts dms(254.6) == %(254°36'00")
puts dms(93.034773) == %(93°02'05")
puts dms(0) == %(0°00'00")
puts dms(360) == %(360°00'00") || dms(360) == %(0°00'00")

# Given solution:

DEGREE = "\xC2\xB0"
MINUTES_PER_DEGREE = 60
SECONDS_PER_MINUTE = 60
SECONDS_PER_DEGREE = MINUTES_PER_DEGREE * SECONDS_PER_MINUTE

def dms(degrees_float)
  total_seconds = (degrees_float * SECONDS_PER_DEGREE).round
  degrees, remaining_seconds = total_seconds.divmod(SECONDS_PER_DEGREE)
  minutes, seconds = remaining_seconds.divmod(SECONDS_PER_MINUTE)
  format(%(#{degrees}#{DEGREE}%02d'%02d"), minutes, seconds)
end

puts dms(30) == %(30°00'00")
puts dms(76.73) == %(76°43'48")
puts dms(254.6) == %(254°36'00")
puts dms(93.034773) == %(93°02'05")
puts dms(0) == %(0°00'00")
puts dms(360) == %(360°00'00") || dms(360) == %(0°00'00")

# Further exploration:

DEGREE = "\xC2\xB0"
MINUTES_PER_DEGREE = 60
SECONDS_PER_MINUTE = 60
SECONDS_PER_DEGREE = MINUTES_PER_DEGREE * SECONDS_PER_MINUTE
MAX_DEGREE = 360

def dms(degrees_float)
  if degrees_float > MAX_DEGREE
    until degrees_float <= MAX_DEGREE
      degrees_float -= MAX_DEGREE
    end
  elsif degrees_float < 0
    until degrees_float > 0
      degrees_float += MAX_DEGREE
    end
  end
  
  # could replace with:
  # degrees_float = degrees_float % MAX_DEGREE

  total_seconds = (degrees_float * SECONDS_PER_DEGREE).round
  degrees, remaining_seconds = total_seconds.divmod(SECONDS_PER_DEGREE)
  minutes, seconds = remaining_seconds.divmod(SECONDS_PER_MINUTE)
  format(%(#{degrees}#{DEGREE}%02d'%02d"), minutes, seconds)
end

puts dms(400) == %(40°00'00")
puts dms(-40) == %(320°00'00")
puts dms(-420) == %(300°00'00")

File: js_210/small_problems/easy_4/exercise_1.js
// Write a function that takes a floating point number representing an angle
// between 0 and 360 degrees and returns a string representing that angle in
// degrees, minutes, and seconds. You should use a degree symbol (˚) to
// represent degrees, a single quote (') to represent minutes, and a double
// quote (") to represent seconds. There are 60 minutes in a degree, and 60
// seconds in a minute.

// Solution

// function format(number) {
//   return String(number).length < 2 ? `0${number}` : String(number);
// }

// function dms(angle) {
//   let degrees = Math.floor(angle);
//   let remainder = angle - degrees;
//   let minutes = Math.floor(remainder * 60);
//   remainder = (remainder * 60) - minutes;
//   let seconds = Math.floor(remainder * 60);
  
//   return `${degrees}°${format(minutes)}'${format(seconds)}"`;
// }

// Given Solution

// const DEGREE = '\xB0';
// const MINUTES_PER_DEGREE = 60;
// const SECONDS_PER_MINUTE = 60;
// const SECONDS_PER_DEGREE = MINUTES_PER_DEGREE * SECONDS_PER_MINUTE;

// function dms(degreesFloat) {
//   const degreesInt = Math.floor(degreesFloat);
//   const minutes = Math.floor((degreesFloat - degreesInt) * MINUTES_PER_DEGREE);
//   const seconds = Math.floor(
//     (degreesFloat - degreesInt - (minutes / MINUTES_PER_DEGREE)) *
//     SECONDS_PER_DEGREE
//   );

//   return `${String(degreesInt) + DEGREE + padZeroes(minutes)}'${padZeroes(seconds)}"`;
// }

// function padZeroes(number) {
//   const numString = String(number);
//   return numString.length < 2 ? (`0${numString}`) : numString;
// }

// console.log(dms(30));           // 30°00'00"
// console.log(dms(76.73));        // 76°43'48"
// console.log(dms(254.6));        // 254°35'59"
// console.log(dms(93.034773));    // 93°02'05"
// console.log(dms(0));            // 0°00'00"
// console.log(dms(360));          // 360°00'00" or 0°00'00"


// Further Exploration

const DEGREE = '\xB0';
const MINUTES_PER_DEGREE = 60;
const SECONDS_PER_MINUTE = 60;
const SECONDS_PER_DEGREE = MINUTES_PER_DEGREE * SECONDS_PER_MINUTE;

const MAX_DEGREES = 360;

function dms(degreesFloat) {
  if (degreesFloat < 0) {
    while (degreesFloat < 0) {
      degreesFloat = MAX_DEGREES + degreesFloat;
    }
  } else if (degreesFloat > MAX_DEGREES) {
    while (degreesFloat > MAX_DEGREES) {
     degreesFloat = degreesFloat - MAX_DEGREES; 
    }
  }

  const degreesInt = Math.floor(degreesFloat);
  const minutes = Math.floor((degreesFloat - degreesInt) * MINUTES_PER_DEGREE);
  const seconds = Math.floor(
    (degreesFloat - degreesInt - (minutes / MINUTES_PER_DEGREE)) *
    SECONDS_PER_DEGREE
  );

  return `${String(degreesInt) + DEGREE + padZeroes(minutes)}'${padZeroes(seconds)}"`;
}

function padZeroes(number) {
  const numString = String(number);
  return numString.length < 2 ? (`0${numString}`) : numString;
}

console.log(dms(-1));   // 359°00'00"
console.log(dms(400));  // 40°00'00"
console.log(dms(-40));  // 320°00'00"
console.log(dms(-420)); // 60°00'00"  ??? 300°00'00" ???
